File: cpp/webcam.cpp
#include "webcam.hpp"

// Constructor
Webcam::Webcam(int device_id) : device_id(device_id), recording(false) {}

// Destructor
Webcam::~Webcam() {
    close();
}

bool Webcam::open() {
    return cap.open(device_id);
}

void Webcam::close() {
    if (recording) {
        stopRecording();
    }

    if (cap.isOpened()) {
        cap.release();
    }
}

bool Webcam::isOpened() const {
    return cap.isOpened();
}

cv::Mat Webcam::captureFrame() {
    if (!cap.isOpened()) {
        throw std::runtime_error("Webcam is not open.");
    }

    cv::Mat frame;
    cap >> frame;
    if (frame.empty()) {
        throw std::runtime_error("Failed to capture frame.");
    }

    // If recording, write the frame to the video file
    if (recording) {
        writer.write(frame);
    }

    return frame;
}

bool Webcam::startRecording(const std::string& filename, int fourcc) {
    if (cap.isOpened()) {
        int width = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_WIDTH));
        int height = static_cast<int>(cap.get(cv::CAP_PROP_FRAME_HEIGHT));

        writer.open(filename, fourcc, 30.0, cv::Size(width, height), true);
        if (!writer.isOpened()) {
            return false;
        }
        recording = true;
        return true;
    }
    return false;
}

void Webcam::stopRecording() {
    if (recording) {
        writer.release();
        recording = false;
    }
}

bool Webcam::isRecording() const {
    return recording;
}

std::string Webcam::getDeviceInfo() const {
    // Expand this with additional info if desired
    return "Webcam ID: " + std::to_string(device_id);
}

File: python/webcam.py
#!/usr/bin/env python3
import cv2

class Webcam:
    def __init__(self, device_id):
        self.device_id = device_id
        self.cap = None
        self.writer = None
        self.recording = False

    def open(self):
        self.cap = cv2.VideoCapture(self.device_id)
        return self.cap.isOpened()

    def close(self):
        if self.cap and self.cap.isOpened():
            self.cap.release()
        if self.recording:
            self.stop_recording()

    def is_opened(self):
        return self.cap.isOpened() if self.cap else False

    def capture_frame(self):
        if not self.is_opened():
            return None
        ret, frame = self.cap.read()
        if ret:
            return frame
        return None

    def start_recording(self, filename):
        if self.is_opened():
            fourcc = cv2.VideoWriter_fourcc(*'MJPG')
            width = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
            height = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
            fps = 30
            self.writer = cv2.VideoWriter(filename, fourcc, fps, (width, height))
            self.recording = True

    def stop_recording(self):
        if self.recording and self.writer:
            self.writer.release()
            self.recording = False

    def get_device_info(self):
        if not self.is_opened():
            return "Device not opened"
        width = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        height = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        fps = self.cap.get(cv2.CAP_PROP_FPS)
        info = f"Width: {width}, Height: {height}, FPS: {fps}"
        return info
